File: cpp/sorting/bucketSort.cpp
#include <vector>
#include <iostream>

using std::vector;
using std::cout;
using std::endl;

// C++ implementation of Bucket Sort
vector<int> bucketSort(vector<int> arr) {
    // Assuming arr only contains 0, 1 or 2
    int counts[] = {0, 0, 0};

    // Count the quantity of each val in arr
    for (int n: arr) {
        counts[n]++;
    }

    int i = 0;
    for (int n = 0; n < 3; n++) {
        for (int j = 0; i < counts[n]; j++) {
            arr[i] = n;
            i++;
        }
    }
    return arr;
}

int main() {
    int a[] = {2, 1, 2, 0, 0, 2};
    //vector<int> arr = vector<int>(a);
    
    bucketSort(arr);
    for (auto n: a) {
        cout << n;
    }
    cout << endl;
    return 0;
}

File: csharp/sorting/BucketSort.cs
public class BucketSort {

    // C# implementation of Bucket Sort
    public static int[] Sort(int[] arr) {

        // Assuming arr only contains 0, 1 or 2
        int[] counts = {0, 0, 0};

        // Count the quantity of each val in arr
        foreach (int num in arr) {
            counts[num] += 1;
        }

        // Fill each bucket in the original array
        int i = 0;
        for (int n = 0; n < counts.Length; n++) {
            for (int j = 0; j < counts[n]; j++) {
                arr[i] = n;
                i++;
            }
        }
        return arr;
    }
}
